From: "Matthew Wilcox (Oracle)" <willy@infradead.org>
Subject: mm: inline destroy_large_folio() into __folio_put_large()
Date: Fri, 5 Apr 2024 16:32:25 +0100

destroy_large_folio() has only one caller, move its contents there.

Link: https://lkml.kernel.org/r/20240405153228.2563754-4-willy@infradead.org
Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Reviewed-by: Zi Yan <ziy@nvidia.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/mm.h |    2 --
 mm/page_alloc.c    |   14 --------------
 mm/swap.c          |   13 ++++++++++---
 3 files changed, 10 insertions(+), 19 deletions(-)

--- a/include/linux/mm.h~mm-inline-destroy_large_folio-into-__folio_put_large
+++ a/include/linux/mm.h
@@ -1318,8 +1318,6 @@ void folio_copy(struct folio *dst, struc
 
 unsigned long nr_free_buffer_pages(void);
 
-void destroy_large_folio(struct folio *folio);
-
 /* Returns the number of bytes in this potentially compound page. */
 static inline unsigned long page_size(struct page *page)
 {
--- a/mm/page_alloc.c~mm-inline-destroy_large_folio-into-__folio_put_large
+++ a/mm/page_alloc.c
@@ -565,20 +565,6 @@ void prep_compound_page(struct page *pag
 	prep_compound_head(page, order);
 }
 
-void destroy_large_folio(struct folio *folio)
-{
-	if (folio_test_hugetlb(folio)) {
-		free_huge_folio(folio);
-		return;
-	}
-
-	if (folio_test_large_rmappable(folio))
-		folio_undo_large_rmappable(folio);
-
-	mem_cgroup_uncharge(folio);
-	free_unref_page(&folio->page, folio_order(folio));
-}
-
 static inline void set_buddy_order(struct page *page, unsigned int order)
 {
 	set_page_private(page, order);
--- a/mm/swap.c~mm-inline-destroy_large_folio-into-__folio_put_large
+++ a/mm/swap.c
@@ -127,9 +127,16 @@ static void __folio_put_large(struct fol
 	 * (it's never listed to any LRU lists) and no memcg routines should
 	 * be called for hugetlb (it has a separate hugetlb_cgroup.)
 	 */
-	if (!folio_test_hugetlb(folio))
-		page_cache_release(folio);
-	destroy_large_folio(folio);
+	if (folio_test_hugetlb(folio)) {
+		free_huge_folio(folio);
+		return;
+	}
+
+	page_cache_release(folio);
+	if (folio_test_large_rmappable(folio))
+		folio_undo_large_rmappable(folio);
+	mem_cgroup_uncharge(folio);
+	free_unref_page(&folio->page, folio_order(folio));
 }
 
 void __folio_put(struct folio *folio)
_
